File: two-pointer/valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        specialChars = set([" ", ",", ".", "!", "?", "'", ":"])

        i = 0
        j = len(s) - 1

        while(i < j):
            while(s[i] in specialChars and i < j):
                i += 1
            while(s[j] in specialChars and i < j):
                j -= 1
            
            left = s[i].lower()
            right = s[j].lower()

            if(left != right):
                return False
            i += 1
            j -= 1
        
        return True

File: two-pointer/valid-palindrome.js
class Solution {
  /**
   * @param {string} s
   * @return {boolean}
   */

  // O(n) time O(1) space
  isPalindrome(s) {
    const specialChars = new Set([" ", ",", ".", "!", "?", "'", ":"]);
    let i = 0;
    let j = s.length - 1;
    while (i < j) {
      while (specialChars.has(s[i]) && i < j) {
        i++;
      }
      while (specialChars.has(s[j]) && i < j) {
        j--;
      }

      let left = s[i].toLowerCase();
      let right = s[j].toLowerCase();
      if (left !== right) {
        console.log({ i, left, j, right });
        return false;
      }
      i++;
      j--;
    }
    return true;
  }
}
